### PR TITLE
V1 of Optimizing Code vs Recognizing Patterns with 3D Arrays

### PR DESCRIPTION
**Pull request by coatless**
Improved [S/O post](http://stackoverflow.com/a/37690098/1345455). Though, it might be a bit wordy now.

Due to the on-off again code blocks, I opted to just copy the results from the initial SO post directly into the `.Rmd`.

### CODE REVIEW

**Comment by eddelbuettel** [comment]
Looks good, only minor question:  do we need a tag 'cube' (which will be a singleton) ?

**Comment by coatless** [comment]
I'm going to say: **Yes**.

Primarily, the resources for multidimensional arrays in `Rcpp` is a bit sparse as there is no supported type.

I toyed around with tagging it `array` but there seems to be [more support](http://www2.cs.uregina.ca/~dbd/cs831/notes/dcubes/dcubes.html) for visualizing this specific form of data via `cube`s.  (Also, `array` has yet to be tagged...)

Personally, I'm going to be doing more with cubes in the upcoming months as I move into multidimensional data sets.

**Comment by eddelbuettel** [comment]
Yes, agreed, I used Cube with Armadillo too.  

Maybe some old posts could use it too.

**Comment by eddelbuettel** [comment]
Random notes now that it is up:
- the assignment in R with `=` is pure evil
- the ending section about your machine will have to go to as it gets re-rendered / re-computed on whichever machines runs it.

**Comment by eddelbuettel** [comment]
One round of editorial oversight applied.  I now understand that you didn't quite write it vignette style with 'fixed' results.  That's fine, if only a wee bit different :)

/cc @jjallaire

**Comment by coatless** [comment]
@eddelbuettel 

That was a quick merge! I'm going to probably submit another PR correcting a few typos and bringing in a dynamic flow to the benchmarks. 

Normally, the piece-by-piece article build is no problem as I can source different components from an external file.

**Comment by coatless** [comment]
On a more serious note though, why is:

> the assignment in R with = is pure evil

?

I'm aware of the:
1.  Different scoping rules that `<-` permits to operate well in function definitions, e.g.
   - `mean(x <- 1:5)` vs. `mean(x=1:5)` the prior enables `x` to live on after the function call while the later does not. 
2. `=` is lower precedence than `<-`
3. Short cuts exist in RStudio to speed up the typing via `Alt` + `-` and `Option` + `-` in RStudio. 
   - However, 3 characters vs. 1 character just seems wrong.

But, I suppose the main reason why I'm particularly perturbed by the `<-` operator is due to a logical comparison. e.g.

`x<-y` potentially causing a logical condition `x < -y` instead of the desired assignment all because of a space.

Thus, the equality symbol just seems safer and more normal. Thus, I normally opt for `=` as a result.

**Comment by eddelbuettel** [comment]
Well `<-` came before `=`, is pretty standard in R code, see eg the base R code itself, and used in all other Gallery posts.   We don't have a formal style guide, but if we did, `<-` would clearly be a strong recommendation.  If it breaks your heart we can revert; I would prefer to keep `<-`.

**Comment by coatless** [comment]
No need to. The strong language against the '=' operator made me slightly worried that I had overlooked something nontrivial.

**Comment by eddelbuettel** [comment]
The initial 'pure evil' was in jest, but there are good reasons to prefer `<-`.  IIRC one of those is the evaluation model as IIRC you can't as easily wrap things like `system.time(....)` etc around `a = foo` assignments.

And then there is style.  `<-` is simply how R does it.
